r2-core/src/test/java/test/r2/transport/http/client/TestAsyncPool.java

```diff
@@ -570,6 +580,389 @@ public void testWaitTimeStats() throws Exception
+    ScheduledExecutorService executor = Executors.newScheduledThreadPool(500);
+    RateLimiter rateLimiter = new ExponentialBackOffRateLimiter(0, 5000,
+        10, executor, concurrency);
+
+    ClockedExecutor clockedExecutor = new ClockedExecutor();
+    SettableClock settableClock = new SettableClock();
+
+    final AsyncPool<Object> pool = new AsyncPoolImpl<Object>("object pool",
+        blockableObjectCreator,
+        poolSize,
+        Integer.MAX_VALUE,
+        waiterTimeout,
+        clockedExecutor,
+        Integer.MAX_VALUE,
+        AsyncPoolImpl.Strategy.MRU,
+        0, rateLimiter, settableClock, new LongTracking()
+    );
+
+    pool.start();
+
+    // Phase A : Checking out object 'numberOfCheckoutsInPhaseA' times !
+    List<Object> checkedOutObjects = performCheckout(numberOfCheckoutsInPhaseA, pool);
+
+    // Phase B : Blocking object creation and performing the checkout 'numberOfCheckoutsInPhaseB' times again
+    blockableObjectCreator.blockCreation();
+    Future<None> future = performUnblockingCheckout(numberOfCheckoutsInPhaseB,
+        0, pool);
+
+    blockableObjectCreator.waitUntilAllBlocked();
+
+    // Phase C : Returning the checkedOut objects from Phase A back to the object pool
+    for (int i = 0; i < numbOfObjectsToBeReturnedInPhaseC; i++)
+    {
+      pool.put(checkedOutObjects.remove(0));
+    }
+
+    settableClock.addDuration(waiterTimeout+1);
+    clockedExecutor.runFor(waiterTimeout+2);
+
+    // Phase D : All the object creation in phase B gets unblocked now
+    blockableObjectCreator.unblockCreation();
+    try
+    {
+      future.get(5, TimeUnit.SECONDS);
+    }
+    catch (Exception e)
+    {
+      Assert.fail("Unexpected interruption", e);
+    }
+
+    // Making sure the rate limiter pending tasks are submitted to the executor
+    while (rateLimiter.numberOfPendingTasks() > 0)
+    {
+      try
+      {
+        Thread.sleep(1);
+      }
+      catch (Exception e)
+      {
+        Assert.fail("Unexpected interruption", e);
+      }
+    }
+
+    executor.shutdown();
+
+    try
+    {
+      if (!executor.awaitTermination(10, TimeUnit.SECONDS))
+      {
+        Assert.fail("Too long to finish");
+      }
+    }
+    catch (Exception ex)
+    {
+      Assert.fail("Unexpected interruption", ex);
+    }
+
+    PoolStats stats = pool.getStats();
+    Assert.assertEquals(stats.getTotalCreationIgnored(), numberOfCheckoutsInPhaseB - concurrency);
+    Assert.assertEquals(stats.getCheckedOut(), numberOfCheckoutsInPhaseA);
+    Assert.assertEquals(stats.getIdleCount(), concurrency);
+    Assert.assertEquals(stats.getTotalCreated(), numberOfCheckoutsInPhaseA + concurrency);
+    Assert.assertEquals(stats.getPoolSize(), numberOfCheckoutsInPhaseA + concurrency);
+    Assert.assertEquals(stats.getTotalWaiterTimedOut(), numberOfCheckoutsInPhaseB - numbOfObjectsToBeReturnedInPhaseC);
+  }
+
+  @DataProvider
+  public Object[][] dataProvider()
```

Comment:
You can specify the name of the dataProvider including what kind of data it'll return